### PR TITLE
support similarity masks

### PR DESCRIPTION
**Pull request by qchateau**
### Content

This PR adds the necessary code to add a similarity mask to the exposure compensator of the stitching pipeline.

You can configure a `similarity_threshold` which defines a threshold on the relative error between the pixels of two images. If the error is above the threshold, it is ignored when calculating the exposure compensation. This allows the removal of ghosting artifacts due to exposure compensation when part of the images are different (for instance a moving car which would only be present on a single picture)

Note that by default the behavior is unchanged.

<cut/>

### Example

The top 2 images are the source images. The bottom left one used a block, per channel exposure compensation. The bottom right used the same, with a similarity threshold of 0.1.

Despite the seam finder and blender doing their job perfectly, the red artifact persists due to the exposure compensator.

The red rectangle has obviously been added artificially but that kind of artifact can exist naturally (example of a moving car above).

Untitled

### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or other license that is incompatible with OpenCV
- [x] The PR is proposed to proper branch
- [x] There is reference to original bug report and related work
- [x] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [x] The feature is well documented and sample code can be built with the project CMake
   -> The feature is not documented, but neither are the other features are the exposure compensators. They belong to the `detail` namespace, I guess this is why ?

### CODE REVIEW

**Comment by alalek** [comment]
Thank you for contribution!

Please add simple test for new feature (reuse existed test code and testdata with adding artificial "rectangle" on input like on example).

**Comment by qchateau** [comment]
Added the test, I hope it's good enough

File: modules/stitching/src/exposure_compensate.cpp [hunk 1]
```diff
@@ -298,6 +308,73 @@ void GainCompensator::setMatGains(std::vector<Mat>& umv)
...
+    for (int i = 0; i < num_images; ++i)
+    {
+        for (int j = i; j < num_images; ++j)
+        {
+            Rect roi;
+            if (overlapRoi(corners[i], corners[j], images[i].size(), images[j].size(), roi))
+            {
+                Mat subimg1 = images[i](Rect(roi.tl() - corners[i], roi.br() - corners[i])).getMat(ACCESS_READ);
+                Mat subimg2 = images[j](Rect(roi.tl() - corners[j], roi.br() - corners[j])).getMat(ACCESS_READ);
+                Mat similarity = getSimilarityMask(subimg1, subimg2);
```
Comment (alalek) [inline modules/stitching/src/exposure_compensate.cpp:336]: 1. Please use `InputArray` for parameters in getSimilarityMask().
2. Move `getMat(ACCESS_READ)` there.
3. It is better to store UMat in `similarities_` field.

File: modules/stitching/src/exposure_compensate.cpp [hunk 2]
```diff
@@ -298,6 +308,73 @@ void GainCompensator::setMatGains(std::vector<Mat>& umv)
...
+    CV_Assert(src1.type() == src2.type());
+    CV_Assert(src1.type() == CV_8UC3 || src1.type() == CV_8UC1);
+
+    Mat similarity(src1.rows, src1.cols, CV_8UC1);
+    for (int y = 0; y < similarity.rows; ++y)
+    {
+        for (int x = 0; x < similarity.cols; ++x)
+        {
+            double diff;
+            if (src1.channels() == 3)
```
Comment (alalek) [inline modules/stitching/src/exposure_compensate.cpp:355]: Avoid constant `if` in per-pixel processing.
Move this check upper.

File: modules/stitching/src/exposure_compensate.cpp [hunk 3]
```diff
@@ -298,6 +308,73 @@ void GainCompensator::setMatGains(std::vector<Mat>& umv)
...
+            {
+                Vec<int, 3> vec_diff =
+                    Vec<int, 3>(*src1.ptr<Vec<uchar, 3>>(y, x))
+                    - Vec<int, 3>(*src2.ptr<Vec<uchar, 3>>(y, x));
+                diff = norm(vec_diff / 255);
```
Comment (alalek) [inline modules/stitching/src/exposure_compensate.cpp:360]: Avoid divisions. Prefer multiplications.

Or move this to `diff <= similarity_threshold_` check (by adjusting threshold).

File: modules/stitching/src/exposure_compensate.cpp [hunk 4]
```diff
@@ -298,6 +308,73 @@ void GainCompensator::setMatGains(std::vector<Mat>& umv)
...
+                Mat subimg1 = images[i](Rect(roi.tl() - corners[i], roi.br() - corners[i])).getMat(ACCESS_READ);
+                Mat subimg2 = images[j](Rect(roi.tl() - corners[j], roi.br() - corners[j])).getMat(ACCESS_READ);
+                Mat similarity = getSimilarityMask(subimg1, subimg2);
+                similarities_.push_back(similarity);
+            }
+        }
+    }
+}
+
+Mat GainCompensator::getSimilarityMask(Mat src1, Mat src2)
```
Comment (alalek) [inline modules/stitching/src/exposure_compensate.cpp:343]: > getSimilarityMask

buildSimilarityMask, prepareSimilarityMask ?

(`getSimilarityMask()` with `similarities_` field may confuse that it just returns field value)

File: modules/stitching/test/test_exposure_compensate.cpp [hunk 5]
```diff
@@ -0,0 +1,113 @@
+/*M///////////////////////////////////////////////////////////////////////////////////////
```
Comment (alalek) [inline modules/stitching/test/test_exposure_compensate.cpp:1]: Please use short license header: https://github.com/opencv/opencv/wiki/Coding_Style_Guide#file-structure

```
// This file is part of OpenCV project.
// It is subject to the license terms in the LICENSE file found in the top-level directory
// of this distribution and at http://opencv.org/license.html.
```

**Review by alalek — COMMENTED** [review]
Great! Thank you for update :+1:

Could you please take a look on comments below.
If you have some issues, then I can prepare proposals in the end of week.

File: modules/stitching/src/exposure_compensate.cpp (hunk 3 above)
Comment (qchateau) [reply modules/stitching/src/exposure_compensate.cpp:360]: Makes me realize this is wrong, it should be `Vec<float, 3>` the code above will (almost) always round down to 0. I'll fix that as well

**Comment by qchateau** [comment]
I ended up reworking `buildSimilarityMask` to follow your advice and fix the bugs. I hope it addresses your comments. Let me know

File: modules/stitching/src/exposure_compensate.cpp [hunk 6]
```diff
@@ -298,6 +308,86 @@ void GainCompensator::setMatGains(std::vector<Mat>& umv)
...
+                    - static_cast<int>(*src2.ptr<uchar>(y, x))) / 255.f;
+
+                *similarity.ptr<uchar>(y, x) = diff <= similarity_threshold_ ? 255 : 0;
+            }
+        }
+    }
+
+    Mat kernel = getStructuringElement(MORPH_RECT, Size(3,3));
+    erode(umat_similarity, umat_similarity, kernel);
+    dilate(umat_similarity, umat_similarity, kernel);
```
Comment (alalek) [inline modules/stitching/src/exposure_compensate.cpp:386]: Inplace arguments are not supported by filtering functions (operation itself is not inplace). Usually data is cloned inside before processing by function implementation.

Please use some temporary array for that.

```
UMat umat_erode;
erode(umat_similarity, umat_erode, kernel);
dilate(umat_erode, umat_similarity, kernel);
```

File: modules/stitching/src/exposure_compensate.cpp [hunk 7]
```diff
@@ -298,6 +308,86 @@ void GainCompensator::setMatGains(std::vector<Mat>& umv)
...
+            for (int x = 0; x < similarity.cols; ++x)
+            {
+                float diff = std::abs(static_cast<int>(*src1.ptr<uchar>(y, x))
+                    - static_cast<int>(*src2.ptr<uchar>(y, x))) / 255.f;
+
+                *similarity.ptr<uchar>(y, x) = diff <= similarity_threshold_ ? 255 : 0;
+            }
+        }
+    }
+
```
Comment (alalek) [inline modules/stitching/src/exposure_compensate.cpp:383]: Please add `similarity.release()` to sync data into `umat_similarity` or use "code scope" in this way

```
{
    Mat similarity = umat_similarity.getMat(ACCESS_WRITE);
    ...
}  // .release() is called automatically

... work with up-to-date umat_similarity ...
```

File: modules/stitching/test/test_exposure_compensate.cpp [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+
+    double psnr = cvtest::PSNR(src1_channels[0], src2_channels[0]);
+    psnr = std::min(psnr, cvtest::PSNR(src1_channels[1], src2_channels[1]));
+    return std::min(psnr, cvtest::PSNR(src1_channels[2], src2_channels[2]));
+}
+
+TEST(ExposureCompensate, SimilarityThreshold)
+{
+  UMat source;
+  imread(cvtest::TS::ptr()->get_data_path() + "stitching/s1.jpg").copyTo(source);
```
Comment (alalek) [inline modules/stitching/test/test_exposure_compensate.cpp:24]: please use 4 spaces for indentation

**Review by alalek — COMMENTED** [review]
Great! Thank you :+1:

**Review by alalek — APPROVED** [review]

**Pull request merged**